src/main/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/app/database/jpa/mapper/induction/FutureWorkInterestsEntityMapper.kt

```diff
@@ -62,6 +62,24 @@ abstract class FutureWorkInterestsEntityMapper {
     dto: UpdateFutureWorkInterestsDto?,
   )
 
+  @ExcludeJpaManagedFieldsIncludingDisplayNameFields
+  @GenerateNewReference
+  @Mapping(target = "createdAtPrison", source = "prisonId")
+  @Mapping(target = "updatedAtPrison", source = "prisonId")
+  @Mapping(target = "interests", ignore = true)
+  abstract fun fromUpdateDtoToEntity(dto: UpdateFutureWorkInterestsDto?): FutureWorkInterestsEntity
+
+  @AfterMapping
+  fun addInterests(dto: UpdateFutureWorkInterestsDto, @MappingTarget entity: FutureWorkInterestsEntity) {
+    val existingTypes = entity.interests?.map { it.key() }
+    dto.interests.forEach {
+      // ensure we only add new interests
+      if (existingTypes == null || !existingTypes.contains(it.key())) {
```

Comment:
I was caught out here initially because this method is called when mapping from `UpdateFutureWorkInterestsDto` to `FutureWorkInterestsEntity`, regardless of whether it's after the new method `fromUpdateDtoToEntity()` immediately above here, or from the existing after `updateEntityFromDto()` one.

Reply:
100% agreed. I wondered the other day when we were talking about it whether MapStruct was the right tool for the job in this case due to the complexity of mapping. But you are right - at least with it working and a great set of tests (specifically the integration tests) we can consider rewriting this if we feel it is the right thing to do 👍